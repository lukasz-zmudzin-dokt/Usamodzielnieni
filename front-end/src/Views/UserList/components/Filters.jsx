import React, { useState } from "react";
import { DEFAULT_INPUT } from "constants/other";
import { Button, Col, Form } from "react-bootstrap";
import FormGroup from "components/FormGroup";
import { userTypes } from "constants/userTypes";
import { userStatuses } from "constants/userStatuses";

const Filters = ({
  setFilter,
  disabled,
  typeList,
  statusList,
  count,
  filters,
}) => {
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const [type, setType] = useState(DEFAULT_INPUT);
  const [status, setStatus] = useState(DEFAULT_INPUT);

  const filter = (e) => {
    e.preventDefault();
    const typeFilter = type !== DEFAULT_INPUT ? type : undefined;
    const statusFilter = status !== DEFAULT_INPUT ? status : undefined;
    const usernameFilter = username !== "" ? username : undefined;
    const emailFilter = email !== "" ? email : undefined;
    setFilter({
      ...filters,
      username: usernameFilter,
      email: emailFilter,
      type: typeFilter,
      status: statusFilter,
    });
  };

  const deleteFilter = () => {
    setUsername("");
    setEmail("");
    setStatus(DEFAULT_INPUT);
    setType(DEFAULT_INPUT);
    setFilter({ ...filters, username: "", email: "", type: "", status: "" });
  };

  return (
    <Form className="search__form" onSubmit={filter}>
      <Form.Row>
        <FormGroup
          as={Col}
          xs={12}
          md={6}
          header="Nazwa użytkownika"
          val={username}
          setVal={setUsername}
          type="text"
          id="username"
        />
        <FormGroup
          as={Col}
          xs={12}
          md={6}
          header="Email"
          val={email}
          setVal={setEmail}
          type="text"
          id="email"
        />
        <Form.Group as={Col} xs={12} md={6} controlId="category">
          <Form.Label>Status:</Form.Label>
          <Form.Control
            as="select"
            type="select"
            value={status}
            onChange={(e) => setStatus(e.target.value)}
          >
            <option disabled>{DEFAULT_INPUT}</option>
            {Object.values(userStatuses).map((status) => (
              <option key={status} value={status}>
                {statusList(status)}
              </option>
            ))}
          </Form.Control>
        </Form.Group>
        <Form.Group as={Col} xs={12} md={6} controlId="type">
          <Form.Label>Typ użytkownika:</Form.Label>
          <Form.Control
            as="select"
            type="select"
            value={type}
            onChange={(e) => setType(e.target.value)}
          >
            <option disabled>{DEFAULT_INPUT}</option>
            {Object.values(userTypes).map((type) => (
              <option key={type} value={type}>
                {typeList(type)}
              </option>
            ))}
          </Form.Control>
        </Form.Group>
      </Form.Row>

      <Button
        type="submit"
        className="mr-3"
        variant="primary"
        disabled={disabled}
      >
        {disabled ? "Ładowanie..." : "Filtruj użytkowników"}
      </Button>
      <Button
        variant="outline-primary"
        className="mr-3"
        onClick={deleteFilter}
        disabled={disabled}
      >
        {disabled ? "Ładowanie..." : "Wyczyść filtry"}
      </Button>
      {count !== 0 && (
        <small className="search__countText">{`Liczba znalezionych użytkowników: ${count}`}</small>
      )}
    </Form>
  );
};

export default Filters;
