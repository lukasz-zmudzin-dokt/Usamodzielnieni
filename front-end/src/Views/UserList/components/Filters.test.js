import {MemoryRouter} from "react-router-dom";
import React from "react";
import Filters from "./Filters";
import {render, fireEvent, waitForElement} from '@testing-library/react';

describe("Filters", () => {
    let setFilter = jest.fn();

    it('should render correctly', () => {
        const {container} = render(
            <MemoryRouter>
                <Filters
                    disabled={false}
                    count={10}
                    setFilter={setFilter}
                    statusList={jest.fn()}
                    typeList={jest.fn()}
                />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    it('should clear filters on click', () => {
        const {getByLabelText, getByText} = render(
            <MemoryRouter>
                <Filters
                    disabled={false}
                    count={10}
                    setFilter={setFilter}
                    statusList={jest.fn()}
                    typeList={jest.fn()}
                />
            </MemoryRouter>
        );
        fireEvent.change(getByLabelText("Nazwa użytkownika", {exact: false}), {
            target: {value: "standard"}
        });
        fireEvent.change(getByLabelText("Email", {exact: false}), {
            target: {value: "qwe@qwe.qwe"}
        });
        fireEvent.change(getByLabelText("Status:", {exact: false}), {
            target: {value: "Verified"}
        });

        fireEvent.click(getByText("Wyczyść filtry"));

        expect(getByLabelText("Nazwa użytkownika").value).toBe("");
        expect(getByLabelText("Email").value).toBe("");
        expect(getByLabelText("Status:").value).toBe("-- Wybierz --");
    });

    it("should not render offers count when offers count is 0", () => {
        const { queryByText } = render(
            <MemoryRouter initialEntries={["/"]}>
                <Filters
                    setFilter={setFilter}
                    count={0}
                    disabled={false}
                    typeList={jest.fn()}
                    statusList={jest.fn()}
                />
            </MemoryRouter>
        );

        expect(queryByText("Liczba znalezionych", { exact: false })).not.toBeInTheDocument();
    });

    it('should call setFilter with right params', () => {
        const { getByLabelText, getByText } = render(
            <MemoryRouter initialEntries={["/"]}>
                <Filters
                    setFilter={setFilter}
                    count={10}
                    disabled={false}
                    typeList={jest.fn()}
                    statusList={jest.fn()}
                />
            </MemoryRouter>
        );

        fireEvent.change(getByLabelText("Nazwa użytkownika", {exact: false}), {
            target: {value: "standard"}
        });
        fireEvent.change(getByLabelText("Email", {exact: false}), {
            target: {value: "qwe@qwe.qwe"}
        });
        fireEvent.change(getByLabelText("Status:", {exact: false}), {
            target: {value: "Verified"}
        });

        fireEvent.click(getByText("Filtruj użytkowników"));

        expect(setFilter).toHaveBeenCalledWith({
            username: "standard",
            email: "qwe@qwe.qwe",
            type: undefined,
            status: "Verified"
        })
    });
});