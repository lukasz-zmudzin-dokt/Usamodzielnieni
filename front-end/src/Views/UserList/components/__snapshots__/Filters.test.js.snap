// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should render correctly 1`] = `
<div>
  <form
    class="search__form"
  >
    <div
      class="form-row"
    >
      <div
        class="form-group col-md-6 col-12"
      >
        <label
          class="form-label"
          for="username"
        >
          Nazwa użytkownika
        </label>
        <input
          class="form-control"
          data-testid="default"
          id="username"
          maxlength="50"
          minlength="1"
          placeholder="Nazwa użytkownika"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group col-md-6 col-12"
      >
        <label
          class="form-label"
          for="email"
        >
          Email
        </label>
        <input
          class="form-control"
          data-testid="default"
          id="email"
          maxlength="50"
          minlength="1"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group col-md-6 col-12"
      >
        <label
          class="form-label"
          for="category"
        >
          Status:
        </label>
        <select
          class="form-control"
          id="category"
          type="select"
        >
          <option
            disabled=""
          >
            -- Wybierz --
          </option>
          <option
            value="verified"
          />
          <option
            value="rejected"
          />
          <option
            value="waiting_for_verification"
          />
          <option
            value="blocked"
          />
        </select>
      </div>
      <div
        class="form-group col-md-6 col-12"
      >
        <label
          class="form-label"
          for="type"
        >
          Typ użytkownika:
        </label>
        <select
          class="form-control"
          id="type"
          type="select"
        >
          <option
            disabled=""
          >
            -- Wybierz --
          </option>
          <option
            value="standard"
          />
          <option
            value="staff"
          />
          <option
            value="employer"
          />
          <option
            value="specialist"
          />
        </select>
      </div>
    </div>
    <button
      class="mr-3 btn btn-primary"
      type="submit"
    >
      Filtruj użytkowników
    </button>
    <button
      class="mr-3 btn btn-outline-primary"
      type="button"
    >
      Wyczyść filtry
    </button>
    <small
      class="search__countText"
    >
      Liczba znalezionych użytkowników: 10
    </small>
  </form>
</div>
`;
